boards/odroid_go: Update partitions size

ydamigos opened the pull request:
Update partitions size to utilize the entire flash.

ydamigos committed:
boards/odroid_go: Fix update partitions size

Update partitions size to utilize the entire flash

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos commented:
@gschwaer Could you please check it?

gschwaer commented:
Hi, @ydamigos, what exactly should I check? I'm not an expert in mcuboot related config, so I'm not sure I can review that well. I can do a test boot on my odroid go if that is what you meant. But then please tell me what exactly I should check, just a boot after flashing? Or also firmware upgrade via second partition? But I'm on vacation the next two weeks. So it'll have to wait until I'm back. Best, Gero

gschwaer commented on boards/hardkernel/odroid_go/odroid_go_appcpu.dts:
```diff
@@ -40,27 +40,27 @@
-			   reg = <0x00210000 0x00040000>;
+			   reg = <0x00A10000 0x00100000>;
 		};
 
```
Maybe you can add a comment like
```
/* Reserve the remaining 5056kB for the storage partition */
```

ydamigos replied:
Added the comment.

marekmatej commented on boards/hardkernel/odroid_go/odroid_go_appcpu.dts:
```diff
@@ -40,27 +40,28 @@
 		};
 
-		/* Reserve 1024kB for the application in slot 0 */
+		/* Reserve 5120kB for the application in slot 0 */
```
Hey, the partitions for the APPCPU should avoid using the same address space the PROCPU is using, therefore current layout doesn't make sense. It's better to leave it as is as it needs to be fixed.

marekmatej reviewed (dismissed):
Approving with a comment on APPCPU layout

ydamigos commented on boards/hardkernel/odroid_go/odroid_go_appcpu.dts:
```diff
@@ -40,27 +40,27 @@
+		storage_partition: partition@B10000 {
 			label = "storage";
-			reg = <0x00250000 0x00006000>;
+			reg = <0x00B10000 0x0004F0000>;
```
Fixed.

ydamigos commented:
> Hi, @ydamigos, what exactly should I check? I'm not an expert in mcuboot related config, so I'm not sure I can review that well. I can do a test boot on my odroid go if that is what you meant. But then please tell me what exactly I should check, just a boot after flashing? Or also firmware upgrade via second partition? But I'm on vacation the next two weeks. So it'll have to wait until I'm back. Best, Gero

Hi @gschwaer,
I just wanted someone else with an odroid go to give a try.

gschwaer commented:
Hi @ydamigos,

unfortunately it doesn't boot on my device. I build the `hello_world` sample on `ydamigos:odroid-go-fix-flash-size`, but it hung. Here's the serial output:

```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x33 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0000,len:10156
ho 0 tail 12 room 4
load:0x40078000,len:6832
load:0x400a0000,len:35840
entry 0x400a2dc8
I (35) boot: MCUboot 2nd stage bootloader
I (35) boot: compile time Aug 21 2024 13:27:00
W (36) boot: Unicore bootloader
I (36) spi_flash: detected chip: generic
I (39) spi_flash: flash io: dio
I (42) boot: chip revision: v1.0
I (45) boot.esp32: SPI Speed      : 40MHz
I (49) boot.esp32: SPI Mode       : DIO
I (53) boot.esp32: SPI Flash Size : 16MB
I (56) boot: Enabling RNG early entropy source...
I (92) spi_flash: flash io: dio
```

As a sanity check I also built ed31037d5fd828ce644f7ca8bbecaf3ef12c1e36 (parent of `ydamigos:odroid-go-fix-flash-size`), that worked:

```
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x33 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0000,len:10156
ho 0 tail 12 room 4
load:0x40078000,len:6832
load:0x400a0000,len:35840
entry 0x400a2dc8
I (35) boot: MCUboot 2nd stage bootloader
I (35) boot: compile time Aug 21 2024 13:24:37
W (36) boot: Unicore bootloader
I (36) spi_flash: detected chip: generic
I (40) spi_flash: flash io: dio
I (42) boot: chip revision: v1.0
I (45) boot.esp32: SPI Speed      : 40MHz
I (49) boot.esp32: SPI Mode       : DIO
I (53) boot.esp32: SPI Flash Size : 16MB
I (56) boot: Enabling RNG early entropy source...
I (92) spi_flash: flash io: dio
[esp32] [INF] Image index: 0, Swap type: none
[esp32] [INF] Loading image 0 - slot 0 from flash, area id: 1
[esp32] [INF] Application start=4008394ch
[esp32] [INF] DRAM segment: paddr=00015aa8h, vaddr=3ffb0000h, size=006ach (  1708) load
[esp32] [INF] IRAM segment: paddr=00010040h, vaddr=40080000h, size=05a68h ( 23144) load
[esp32] [INF] DROM segment: paddr=00020000h, vaddr=3f400000h, size=0109Ch (  4252) map
[esp32] [INF] IROM segment: paddr=00030000h, vaddr=400d0000h, size=03504h ( 13572) map

I (142) quad_psram: This chip is ESP32-D0WD
I (142) esp_psram: Found 4MB PSRAM device
I (142) esp_psram: Speed: 40MHz
I (142) esp_psram: PSRAM initialized, cache is in normal (1-core) mode.
I (994) esp_psram: SPI SRAM memory test OK
*** Booting Zephyr OS build v1.7.99-87044-ged31037d5fd8 ***
Hello World! odroid_go/esp32/procpu
```

I built, flashed and tested with 
```
west build --sysbuild -p always -b odroid_go/esp32/procpu zephyr/samples/hello_world
west flash
# script resets the chip when connecting to serial
tio --script "set{DTR=high}; msleep(100); set{DTR=low}" /dev/serial/by-id/usb-Silicon_Labs_CP2104_USB_to_UART_Bridge_Controller_*-if00-port0
```

Flashing log is identical for both cases:
<details>
<summary>Log</summary>

```text
-- west flash: rebuilding
[0/6] Performing build step for 'hello_world'
ninja: no work to do.
[1/6] Performing build step for 'mcuboot'
ninja: no work to do.
[6/6] Completed 'mcuboot'
-- west flash: using runner esp32
-- runners.esp32: reset after flashing requested
-- runners.esp32: Flashing esp32 chip on None (921600bps)
esptool.py v4.7.0
Found 2 serial ports
Serial port /dev/ttyUSB0
Connecting....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting....
Detecting chip type... ESP32
Chip is ESP32-D0WDQ6 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, BLK3 partially reserved, Coding Scheme 3/4
Crystal is 40MHz
MAC: b4:e6:2d:ae:5b:b9
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 16MB
Flash will be erased from 0x00001000 to 0x0000dfff...
SHA digest in image updated
Wrote 65536 bytes at 0x00001000 in 1.1 seconds (456.4 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
-- west flash: using runner esp32
-- runners.esp32: reset after flashing requested
-- runners.esp32: Flashing esp32 chip on None (921600bps)
esptool.py v4.7.0
Found 2 serial ports
Serial port /dev/ttyUSB0
Connecting.....
Detecting chip type... Unsupported detection protocol, switching and trying again...
Connecting......
Detecting chip type... ESP32
Chip is ESP32-D0WDQ6 (revision v1.0)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, BLK3 partially reserved, Coding Scheme 3/4
Crystal is 40MHz
MAC: b4:e6:2d:ae:5b:b9
Uploading stub...
Running stub...
Stub running...
Changing baud rate to 921600
Changed.
Configuring flash size...
Auto-detected Flash size: 16MB
Flash will be erased from 0x00010000 to 0x00033fff...
Wrote 147456 bytes at 0x00010000 in 2.4 seconds (499.1 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
```

</details> 

If I should try try something else, let me know.

Best,
Gero

ydamigos commented:
Hi @gschwaer,

Thanks a lot for testing it. Truth be told I haven't tested it with `sysbuild`.  Does it work if you don't use `--sysbuild`?

ydamigos commented:
The restriction comes from `CONFIG_BOOT_MAX_IMG_SECTORS=512` defined in `bootloader/mcuboot/boot/zephyr/socs/esp32_procpu.conf`.
It allows maximum image size 2MB. I tried to increase to 1280 but I get an overflow while linking mcuboot `region dram_seg overflowed by 4392 bytes`.

As I am not that familiar with mcuboot, I see two options here, use maximum image size 2 MB and give the rest to scratch and storage or close the PR.

I will go with the first one.

gschwaer reviewed (approved):
I think it's a worth while increment. Before, we didn't use the full flash, now we use it, though with "only" double the image size. Making this work with arbitrary image sizes would be nice, but can stay future work.

I tested this on my device and it booted fine.

ydamigos commented:
@gschwaer Thanks for checking the PR.

ydamigos commented:
@sylvioalves Could you take a look?

marekmatej commented on boards/hardkernel/odroid_go/odroid_go_appcpu.dts:
```diff
@@ -40,27 +40,28 @@
 
-		/* Reserve 1024kB for the application in slot 1 */
-		slot1_partition: partition@110000 {
+		/* Reserve 2048kB for the application in slot 1 */
```
you don't need a scratch partition for the MCUboot.

ydamigos replied:
Scratch partition removed.

marekmatej replied:
@ydamigos, I should be more specific. While the scratch partition is no longer necessary for the MCUboot image manipulations, it is still used in some tests to store data. Size of 64kB should be more than enough.

ydamigos replied:
Done

marekmatej commented:
LGTM - Maybe a nit, why not take the flash partition table out and place it in a separate file to prevent code copy?

ydamigos committed:
odroid_go: Move flash partition table to a separate file

Move flash partition table to a separate file

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos committed:
odroid_go: Enable CONFIG_ESP32_USE_UNSUPPORTED_REVISION

odroid_go is using ESP32 chip revision (1) that is unsupported,
so we need to enable 'CONFIG_ESP32_USE_UNSUPPORTED_REVISION'

Signed-off-by: Yannis Damigos <giannis.damigos@gmail.com>

ydamigos commented:
> LGTM - Maybe a nit, why not take the flash partition table out and place it in a separate file to prevent code copy?

Done. I also enabled 'CONFIG_ESP32_USE_UNSUPPORTED_REVISION' as odroid_go is using ESP32 chip revision (1) that is unsupported.

sylvioalves commented on boards/hardkernel/odroid_go/odroid_go_procpu_defconfig:
```diff
@@ -10,3 +10,5 @@ CONFIG_GPIO=y
 # required to enable LCD backlight
 CONFIG_REGULATOR=y
+
+CONFIG_ESP32_USE_UNSUPPORTED_REVISION=y
```
Would you mind fixing this extra line?
